### PR TITLE
Feature/15 element

### LINKED ISSUE #13: Implement ElementFinder
- [x] ElementFinder

- [x] RelativeElementFinder

### PR DESCRIPTION
**Pull request by knysh**
##### Related Issue Link: 

closes #13 

##### How Has This Been Tested

<class name="tests.elements.WebElementTests"/>
<class name="tests.elements.WindowsElementTests"/>

##### Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: src/main/java/aquality/selenium/core/elements/Element.java [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+        }
+
+        return elementCacheHandler;
+    }
+
+    protected Logger getLogger() {
+        return Logger.getInstance();
+    }
+
+    protected ElementState getElementState() {
```
Comment (mialeska) [inline src/main/java/aquality/selenium/core/elements/Element.java]: do we need it to be public? we don't have it in dotnet

File: src/test/java/tests/elements/factory/CustomElementFactory.java [hunk 2]
```diff
@@ -18,6 +18,7 @@
     protected Map<Class<? extends IElement>, Class<? extends IElement>> getElementTypesMap() {
         Map<Class<? extends IElement>, Class<? extends IElement>> typesMap = new HashMap<>();
         typesMap.put(ICustomElement.class, CustomElement.class);
+        typesMap.put(IWebCustomElement.class, CustomWebElement.class);
```
Comment (mialeska) [inline src/test/java/tests/elements/factory/CustomElementFactory.java]: do you need it? do you check that the element could be got using the interface?

File: src/main/java/aquality/selenium/core/elements/Element.java (hunk 1 above)
Comment (knysh) [reply src/main/java/aquality/selenium/core/elements/Element.java]: it was in selenium-java
I am not sure, lets remove it

File: src/test/java/tests/elements/factory/CustomElementFactory.java (hunk 2 above)
Comment (knysh) [reply src/test/java/tests/elements/factory/CustomElementFactory.java]: added mistakenly, removed

File: src/test/java/tests/elements/factory/ElementFactoryTests.java [hunk 3]
```diff
@@ -148,51 +148,50 @@ public void shouldThrowIllegalArgumentExceptionInGetCustomElementElementTypeIsUn
...
 
     @Test
     public void shouldSetCorrectParametersWhenGettingElement() {
         String name = "1some2";
-        CustomElement element = defaultFactory().getCustomElement(CustomElement::new, CalculatorWindow.getEqualsButton(), name, ElementState.EXISTS_IN_ANY_STATE);
-        Assert.assertEquals(element.getLocator(), CalculatorWindow.getEqualsButton());
+        CustomElement element = defaultFactory().getCustomElement(CustomElement::new, CalculatorWindow.getEqualsButtonLoc(), name, ElementState.EXISTS_IN_ANY_STATE);
+        Assert.assertEquals(element.getLocator(), CalculatorWindow.getEqualsButtonLoc());
         Assert.assertEquals(element.getName(), name);
-        Assert.assertEquals(element.getState(), ElementState.EXISTS_IN_ANY_STATE);
```
Comment (mialeska) [inline src/test/java/tests/elements/factory/ElementFactoryTests.java]: oh no, I believe we still need to test the state. please restore the getState() method in test classes

File: src/test/java/tests/elements/factory/CustomElement.java [hunk 4]
```diff
@@ -1,50 +1,60 @@
...
     }
 
     @Override
-    public <T extends IElement> T findChildElement(By childLoc, IElementSupplier<T> supplier, ElementState state) {
-        throw new NotImplementedException("not implemented in tests");
+    protected IConditionalWait getConditionalWait() {
+        return AqualityServices.get(IConditionalWait.class);
     }
 
-    public ElementState getState() {
```
Comment (mialeska) [inline src/test/java/tests/elements/factory/CustomElement.java]: we still need it I believe

File: src/test/java/tests/elements/factory/ElementFactoryTests.java (hunk 3 above)
Comment (mialeska) [reply src/test/java/tests/elements/factory/ElementFactoryTests.java]: or could be checked via reflection ofc

Comment (knysh) [reply src/test/java/tests/elements/factory/ElementFactoryTests.java]: check, but it wasn't possible

File: src/test/java/tests/elements/factory/CustomElement.java (hunk 4 above)
Comment (knysh) [reply src/test/java/tests/elements/factory/CustomElement.java]: fixed

File: src/test/java/tests/elements/factory/ElementFactoryTests.java (hunk 3 above)
Comment (knysh) [reply src/test/java/tests/elements/factory/ElementFactoryTests.java]: fixed

File: src/main/java/aquality/selenium/core/elements/interfaces/IParent.java [hunk 5]
```diff
@@ -8,31 +8,100 @@
...
-     * @param state visibility state of target element
+     * @param childLoc child element locator
+     * @param clazz    class or interface of the element to be obtained
+     * @param state    visibility state of target element
+     * @param <T>      the type of the element to be obtained
      * @return found child element
      */
-    <T extends IElement> T findChildElement(By childLoc, Class<? extends IElement> clazz, ElementState state);
+    default <T extends IElement> T findChildElement(By childLoc, Class<T> clazz, ElementState state) {
+        return findChildElement(childLoc, null, clazz, state);
```
Comment (DmitryBogatko) [inline src/main/java/aquality/selenium/core/elements/interfaces/IParent.java]: I think we can forbid to create elements without names, can't we? 
So, I suggest removing such signatures 
here and in the method below

File: src/main/resources/localization/be.json [hunk 6]
```diff
@@ -8,5 +8,6 @@
   "loc.elements.were.found.but.not.in.state": "Знайшлі элементы па лакатару '%1$s', але яны не ў жаданым стане %2$s",
   "loc.elements.found.but.should.not": "Не павінна быць знойдзена элементаў па лакатару '%1$s' у %2$s стане",
   "loc.search.of.elements.failed": "Пошук элемента па лакатару '%1$s' прайшоў няўдала",
-  "loc.element.not.in.state": "Элемент %1$s ня стаў %2$s пасля таймаўта %3$s"
+  "loc.element.not.in.state": "Элемент %1$s ня стаў %2$s пасля таймаўта %3$s",
```
Comment (DmitryBogatko) [inline src/main/resources/localization/be.json]: ня - не ?  :) to correspond the school programm

Comment (knysh) [reply src/main/resources/localization/be.json]: fixed

File: src/main/java/aquality/selenium/core/elements/interfaces/IParent.java (hunk 5 above)
Comment (knysh) [reply src/main/java/aquality/selenium/core/elements/interfaces/IParent.java]: discussed

**Pull request merged**